File: internal/scraper/cgmood.go
package scraper

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/struffel/3d-assets-one/internal/database"
	"github.com/struffel/3d-assets-one/internal/model"
)

type CgMoodScraper struct {
	DB *sql.DB
}

func (s *CgMoodScraper) Creator() model.Creator { return model.CreatorCGMood }

var cgMoodURLTypeRegex = map[string]model.AssetType{
	"/3d-model/": model.AssetTypeModel3D,
	"/material/": model.AssetTypePBRMaterial,
}

func (s *CgMoodScraper) ScrapeAssets(ctx context.Context, existing []model.StoredAsset) ([]ScrapedAsset, error) {
	const maxPages = 3
	const indexBase = "https://cgmood.com/free?page="

	tagRe := regexp.MustCompile(`[^A-Za-z0-9]`)

	pageStr, _ := database.GetCreatorState(s.DB, int(model.CreatorCGMood), "page")
	page, _ := strconv.Atoi(pageStr)
	if page < 1 {
		page = 1
	}

	var results []ScrapedAsset
	pagesProcessed := 0

	for pagesProcessed < maxPages {
		body, _, err := FetchURL(ctx, fmt.Sprintf("%s%d", indexBase, page))
		if err != nil {
			break
		}

		doc, err := goquery.NewDocumentFromReader(strings.NewReader(body))
		if err != nil {
			break
		}

		imgElements := doc.Find(".product img")
		if imgElements.Length() == 0 {
			page = 1
			break
		}

		imgElements.Each(func(_ int, sel *goquery.Selection) {
			productURL, _ := sel.Attr("data-product-url")
			productTitle, _ := sel.Attr("data-product-title")
			imgSrc, _ := sel.Attr("src")

			// Determine type from URL
			var at model.AssetType
			found := false
			for pattern, t := range cgMoodURLTypeRegex {
				if strings.Contains(productURL, pattern) {
					at = t
					found = true
					break
				}
			}
			if !found || ContainsURL(existing, productURL) {
				return
			}

			// Tags from title
			parts := tagRe.Split(productTitle, -1)
			var tags []string
			for _, p := range parts {
				if p != "" {
					tags = append(tags, p)
				}
			}

			thumbURL := "https://cgmood.com" + imgSrc
			img, err := FetchImage(ctx, thumbURL)
			if err != nil {
				slog.Warn("cgmood: thumbnail failed", "url", productURL, "err", err)
				return
			}

			results = append(results, ScrapedAsset{
				Title:        productTitle,
				URL:          productURL,
				Tags:         tags,
				Type:         at,
				Creator:      model.CreatorCGMood,
				Status:       model.ScrapedNewlyFound,
				RawThumbnail: img,
			})
		})

		page++
		pagesProcessed++
	}

	database.SetCreatorState(s.DB, int(model.CreatorCGMood), "page", strconv.Itoa(page))
	return results, nil
}

func (s *CgMoodScraper) ValidateAsset(ctx context.Context, asset model.StoredAsset) bool {
	body, status, err := FetchURL(ctx, asset.URL)
	if err != nil || status != 200 {
		return false
	}
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(body))
	if err != nil {
		return false
	}
	btn := doc.Find(".download-button")
	if btn.Length() == 0 {
		return false
	}
	text := btn.First().Text()
	return strings.Contains(text, "Free download")
}

File: app/creator/logic/cgmood.py
import re

from app.asset.asset_type import AssetType
from app.asset.scraped_asset import ScrapedAsset
from app.asset.scraped_asset_collection import ScrapedAssetCollection
from app.asset.scraped_asset_status import ScrapedAssetStatus
from app.asset.stored_asset import StoredAsset
from app.asset.stored_asset_collection import StoredAssetCollection
from app.creator.creator import Creator
from app.creator.creator_logic import CreatorLogic
from app.fetch.web_item_reference import WebItemReference
from app.log.log import Log
from app.log.log_level import LogLevel


class CreatorLogicCgMood(CreatorLogic):
    creator = Creator.CGMOOD
    max_assets_per_run = 3
    indexing_base_url = "https://cgmood.com/free?page="
    url_type_patterns = {
        r"/3d-model/": AssetType.MODEL_3D,
        r"/material/": AssetType.PBR_MATERIAL,
    }

    def validate_asset(self, asset: StoredAsset) -> bool:
        response = WebItemReference(asset.url).fetch()
        if response.http_status_code != 200:
            return False
        dom = response.parse_as_dom()
        if dom is None:
            return False
        download_buttons = dom.find_all(class_="download-button")
        if not download_buttons:
            return False
        button_text = download_buttons[0].get_text()
        return bool(re.search(r"Free download", button_text))

    def scrape_assets(self, existing_assets: StoredAssetCollection) -> ScrapedAssetCollection:
        tmp_collection = ScrapedAssetCollection()
        page = int(self.get_creator_state("page") or 1)
        pages_processed = 0

        while pages_processed < self.max_assets_per_run:
            dom = WebItemReference(self.indexing_base_url + str(page)).fetch().parse_as_dom()
            if dom is None:
                break

            asset_images = dom.find_all("img", attrs={"data-product-url": True})

            for img in asset_images:
                product_url = img.get("data-product-url", "")
                asset_type: AssetType | None = None
                for pattern, t in self.url_type_patterns.items():
                    if re.search(pattern, product_url):
                        asset_type = t

                if not asset_type:
                    Log.write("Skipping URL (no type match)", product_url, LogLevel.WARNING)
                    continue

                if existing_assets.contains_url(product_url):
                    continue

                title = img.get("data-product-title", "")
                tags = [t for t in re.split(r"[^A-Za-z0-9]", title) if t]

                thumbnail_src = img.get("src", "")
                thumbnail_url = "https://cgmood.com" + thumbnail_src if thumbnail_src.startswith("/") else thumbnail_src

                asset = ScrapedAsset(
                    id=None,
                    creator_given_id=None,
                    title=title,
                    url=product_url,
                    tags=tags,
                    type=asset_type,
                    creator=Creator.CGMOOD,
                    status=ScrapedAssetStatus.NEWLY_FOUND,
                    raw_thumbnail=WebItemReference(url=thumbnail_url).fetch().parse_as_image(),
                )
                tmp_collection.append(asset)

            page += 1
            pages_processed += 1

            if not asset_images:
                page = 1

        self.set_creator_state("page", page)
        return tmp_collection
